File: Leetcode/python/Longest Substring Without Repeating Characters.py
class Solution:
    # @return an integer
    def lengthOfLongestSubstring(self, s):
        st = set()
        (p, q) = (0, 0)
        ans = 0
        for c in s:
            if c not in st:
                st.add(c)
                q += 1
                ans = max(ans, q - p)
            else:
                while p < q and c in st:
                    st.remove(s[p])
                    p += 1
                st.add(c)
                q += 1
        return ans
            

File: Leetcode/java/Longest Substring Without Repeating Characters.java
public class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        if (n == 0) {
            return 0;
        }
        HashSet<Character> hs = new HashSet<Character>();
        int p = 0, q = 0;
        int ans = 0;
        while (q < n) {
            while (p < q && hs.contains(s.charAt(q))) {
                hs.remove(s.charAt(p));
                p++;
            }
            hs.add(s.charAt(q));
            q++;
            ans = Math.max(ans, q - p);
        }
        return ans;
    }
}